src/isl.rs

```diff
@@ -0,0 +1,211 @@
+                            "float" => IslTypeRef::IslCoreType(IonType::Float),
+                            "decimal" => IslTypeRef::IslCoreType(IonType::Decimal),
+                            "timestamp" => IslTypeRef::IslCoreType(IonType::Timestamp),
+                            "string" => IslTypeRef::IslCoreType(IonType::String),
+                            "symbol" => IslTypeRef::IslCoreType(IonType::Symbol),
+                            "bool" => IslTypeRef::IslCoreType(IonType::Boolean),
+                            "blob" => IslTypeRef::IslCoreType(IonType::Blob),
+                            "clob" => IslTypeRef::IslCoreType(IonType::Clob),
+                            "sexp" => IslTypeRef::IslCoreType(IonType::SExpression),
+                            "list" => IslTypeRef::IslCoreType(IonType::List),
+                            "struct" => IslTypeRef::IslCoreType(IonType::Struct),
+                            // TODO: add a match for other core types like: lob, text, number, document, any
+                            _ => IslTypeRef::AliasType(type_reference.to_owned()),
+                        };
+                        Ok(ion_type)
+                    })
+            }
+            IonType::Struct =>
+                Ok(IslTypeRef::AnonymousType(value.as_struct().unwrap().try_into()?)),
+            _ => Err(invalid_schema_error_raw(
+                "type reference can either be a symbol(For base/alias type reference) or a struct (for anonymous type reference)",
+            )),
+        }
+    }
+
+    /// Resolves a type_reference into a [TypeId] using the type_store
+    pub fn resolve_type_reference(
+        type_reference: &IslTypeRef,
+        type_store: &SharedTypeStore,
+    ) -> IonSchemaResult<TypeId> {
+        match type_reference {
+            IslTypeRef::IslCoreType(ion_type) => {
+                // TODO: create CoreType struct for storing ISLCoreType type definition instead of Type
+                // inserts ISLCoreType as a Type into type_store
+                Ok(type_store.borrow_mut().add_named_type(
+                    &format!("{:?}", ion_type),
+                    TypeDefinition::new(Some(format!("{:?}", ion_type)), vec![]),
+                ))
+            }
+            IslTypeRef::AliasType(alias) => {
+                let borrow_type_store = type_store.borrow_mut();
+                // verify if the AliasType actually exists in the type_store or throw an error
+                match borrow_type_store.get_type_id_by_name(alias) {
+                    Some(type_id) => Ok(type_id.to_owned()),
+                    None => match borrow_type_store.get_parent() {
+                        Some(parent) => {
+                            // if it is a self referencing type resolve it using parent information from type_store
+                            if parent.0.eq(alias) {
+                                Ok(parent.1)
+                            } else {
```

Comment:
TODO: We'll need to generalize this to work at any level of nesting, not just one.

Reply:
As per offline, current implementation will work fine as the named type can be defined only as a top level type, it should be sufficient to just have a parent which stores the named type's name and `TypeId` rather than storing a `Vec` of parent types for any level of nesting. The named type is going to be the only parent with a name and hence any self reference type can refer to this named type parent to resolve the type. I will also add more doc comments around the usage of parent in `TypeStore`.